### PR TITLE
Add DecimalPrecision type

### PR DESCRIPTION
**Pull request by KaiWitt**
Related: https://github.com/RGB-WG/rgb-core/pull/88

### CODE REVIEW

File: src/fungible/amount.rs [hunk 1]
```diff
@@ -25,7 +25,8 @@
 //! floating point amount with `*_amount` suffix, and reserve `*_value` suffix
 //! for methods returning atomic value.
 
-use std::ops::{Add, AddAssign};
+use std::convert::TryFrom;
```
Comment (dr-orlovsky) [inline src/fungible/amount.rs:28]: We are already on edition 2021, where importing `TryFrom` is not required anymore

File: src/fungible/amount.rs [hunk 2]
```diff
@@ -141,3 +142,86 @@ impl AddAssign for PreciseAmount {
         }
     }
 }
+
+#[derive(Clone, Debug, Display, PartialEq, Eq)]
+#[display(doc_comments)]
+/// Asset decimal precision exceeds 18 digits
+pub struct DecimalPrecisionOverflowError;
```
Comment (dr-orlovsky) [inline src/fungible/amount.rs:149]: Rust API guidelines suggest not to add `Error` suffix to names which are obvious for being errors (like containing `Overflow`, `Invalid`) etc)

File: src/fungible/amount.rs [hunk 3]
```diff
@@ -141,3 +142,86 @@ impl AddAssign for PreciseAmount {
         }
     }
 }
+
+#[derive(Clone, Debug, Display, PartialEq, Eq)]
```
Comment (dr-orlovsky) [inline src/fungible/amount.rs:146]: We can also make it copy

File: src/fungible/amount.rs [hunk 4]
```diff
@@ -141,3 +142,86 @@ impl AddAssign for PreciseAmount {
...
+}
+
+impl TryFrom<usize> for DecimalPrecision {
+    type Error = DecimalPrecisionOverflowError;
+
+    fn try_from(value: usize) -> Result<Self, Self::Error> {
+        Self::try_from(u8::try_from(value).map_err(|_| DecimalPrecisionOverflowError)?)
+    }
+}
+
```
Comment (dr-orlovsky) [inline src/fungible/amount.rs:188]: It would be nice to have convenience constructors like `::zero()`, `::one()`.

**Review by dr-orlovsky — APPROVED** [review]
Thank you very much! I think I can merge it as it is, but I left few nits if you don't mind to add another improvements

File: src/fungible/amount.rs [hunk 5]
```diff
@@ -141,3 +142,86 @@ impl AddAssign for PreciseAmount {
...
+    #[test]
+    fn decimal_precision() {
+        // Minimal
+        assert_eq!(*DecimalPrecision::try_from(u5::ZERO).unwrap(), 0);
+        assert_eq!(*DecimalPrecision::try_from(0u8).unwrap(), 0);
+        assert_eq!(*DecimalPrecision::try_from(0usize).unwrap(), 0);
+
+        // Maximal
+        let u5_18 = u5::try_from(18).unwrap();
```
[the remark this answers is not in the document]
Comment (KaiWitt) [reply src/fungible/amount.rs:208]: I'm sorry I'm not sure what you mean😅
In the `try_from` it's defined as `0..=18` and the test shows that 18 does work

[the remark this answers is not in the document]
Comment (dr-orlovsky) [reply src/fungible/amount.rs:208]: Yes, I just was not sure what is the highest possible precision for u64 type: 17 digits or 18 digits )

**Review by dr-orlovsky — APPROVED** [review]
ACK d69152c6613884dcbed77622c566917f8748b3c1

**Pull request merged**